### PR TITLE
Set AD rules

### LINKED ISSUE #92: AD rules that apply to KroneckerProducts
(Related to #11)

I'm trying to wrap my head around getting gradients with `kron`/`kronecker`.

1. Is it sufficient to define custom AD rules for the vec-trick with [ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl)

```julia
function rrule(::typeof(*), K::KroneckerProduct, x::AbstractVector)
    function times_vec_pullback(ΔΩ)
        ...
    end
    return K*x, times_vec_pullback
end

function rrule(::typeof(*), K::KroneckerProduct, X::AbstractMatrix)
    function times_mat_pullback(ΔΩ)
        ...
    end
    return K*X, times_mat_pullback
end
```

2.  Do we also need to define rules for the constructor as well to get gradients?

```julia
function rrule(::typeof(kronecker), A::AbstractMatrix, B::AbstractMatrix)
    function kronecker_pullback(ΔΩ)
        ...
    end
    return kronecker(A, B), kronecker_pullback
end
```

3. Should the pullbacks also be lazy? I found this to be a [decent overview](https://towardsdatascience.com/shab-dev-story-01-trying-to-find-the-matrix-form-of-gradient-descent-via-backprop-1c2bf7fdf5fe) on finding vectorized derivatives. Would the pullbacks then just be reshape rules for these vectorized derivatives?

### PR DESCRIPTION
**Pull request by elisno**
Resolves #92.

### CODE REVIEW

**Comment by MichielStock** [comment]
Your code does not seem to work for some examples and gives the wrong result for others.

For example:
```julia
gradient((A, B)->sum(A⊗B), A, B)
gradient((A, B)->sum(kron(A,B)), A, B)
```

Most of our `Kronecker` functions fall back on regular function Zygote etc should be able to handle fine. It works for `tr` but not for `logdet` and `sum` (which work with the native `kron`, e.g. `gradient((A, B)-> sum(kron(A, B)), A, B)`. Not sure why or how we can make `ChainRulesCore` fall back on this underlying code.

Do you have a reference for your gradients?

**Comment by elisno** [comment]
> Your code does not seem to work for some examples and gives the wrong result for others.

You're right, I started with the following loss function:

```julia
function loss(A, B, X)
    Z = kron(A, B)*X - y
    L = 0.5 * tr(Z' * Z)
    return L
end
```
where `y` has size `(1, num_samples)`.
I wrote `kronecker_product_pullback` in the `rrule` with this in mind, but forgot that each sample in `y` can have a higher dimension. 

In `test/testchainrules.jl`, I make a comparison of `Zygote.gradient` with hand-written gradients for this trivial case. I do another comparison with `kronecker`.

I decided to leave similar tests for higher-dimensions, but leave them with `@test_broken` for now.

**Comment by elisno** [comment]
I've been experimenting with `KroneckerSum` as well. 

I managed to get the correct values for the pullback:
```julia
function ChainRulesCore.frule((_, ΔA, ΔB), ::KroneckerSum, A::AbstractMatrix, B::AbstractMatrix)
    Ω = (A ⊕ B)
    ∂Ω = (ΔA ⊕ ΔB)
    return Ω, ∂Ω
end

function ChainRulesCore.rrule(::typeof(KroneckerSum), A::AbstractMatrix, B::AbstractMatrix)
    function kronecker_sum_pullback(ΔΩ)
        ∂A = nB .* A + Diagonal(fill(tr(B), nA))
        ∂B = nA .* B + Diagonal(fill(tr(A), nB))
        return (NO_FIELDS, ∂A, ∂B)
    end
    return (A ⊕ B), kronecker_sum_pullback
end

nA = 3
nB = 2
Ar = rand(nA,nA)
Br = rand(nB,nB)
Y_lazy, back_lazy = Zygote._pullback(⊕, Ar, Br)
Y, back = Zygote._pullback((x,y) -> kron(x, Diagonal(ones(nB))) + kron(Diagonal(ones(nA)), y), Ar, Br)
```

```julia
julia> back(Y)[2:end] .≈ back_lazy(Y_lazy)[2:end]
(true, true)
```

**Comment by elisno** [comment]
Note that:
```julia
ChainRulesCore.rrule(::typeof(KroneckerSum), A::AbstractMatrix, B::AbstractMatrix)
```
overwrites
```julia
ChainRulesCore.rrule(::typeof(KroneckerProduct), A::AbstractMatrix, B::AbstractMatrix)
```

Should I use something else instead of `::typeof(KroneckerProduct)`/`::typeof(KroneckerSum)`?

**Comment by MichielStock** [comment]
Still stuck on this, why does computing gradients work for `logdet` but not `tr` or `sum`. It should just fall back to the simple shortcuts, for which adjoints already exist?

**Comment by MichielStock** [comment]
Technically, it only makes sense to define the adjoints for those function where Kronecker provides shortcuts, based on this rule: https://en.wikipedia.org/wiki/Matrix_calculus#Identities_in_differential_form

**Comment by elisno** [comment]
> Still stuck on this, why does computing gradients work for `logdet`

Can you provide a MWE for `logdet`?

**Comment by elisno** [comment]
> Technically, it only makes sense to define the adjoints for those function where Kronecker provides shortcuts, based on this rule: https://en.wikipedia.org/wiki/Matrix_calculus#Identities_in_differential_form

Maybe I misunderstood, but doesn't this only provide the `frule`s?
